OsmAnd/src/net/osmand/plus/views/Renderable.java

```diff
@@ -0,0 +1,480 @@
+                if (culled == null || hash != hashCode) {
+                    if (culler != null)
+                        culler.cancel(true);                // stop any still-running cull
+                    hash = hashCode;
+                    double cullDistance = Math.pow(2.0,param1-zoom);
+                    culler = new AsynchronousResampler(renderType, view, this, cullDistance, param2);
+                    culled = null;                // effectively use full-resolution until re-cull complete (see [Note 1] below)
+                    culler.execute("");
+                }
+            }
+        }
+
+
+        public void drawSingleSegment(Paint p, Canvas canvas, RotatedTileBox tileBox) {
+
+            List<GPXUtilities.WptPt> pts = culled == null? points: culled;			// [Note 1]: use culled points preferentially
+
+            final QuadRect latLonBounds = tileBox.getLatLonBounds();
+
+            int startIndex = -1;
+            int endIndex = -1;
+            int prevCross = 0;
+            double shift = 0;
+            for (int i = 0; i < pts.size(); i++) {
+                GPXUtilities.WptPt ls = pts.get(i);
+                int cross = 0;
+                cross |= (ls.lon < latLonBounds.left - shift ? 1 : 0);
+                cross |= (ls.lon > latLonBounds.right + shift ? 2 : 0);
+                cross |= (ls.lat > latLonBounds.top + shift ? 4 : 0);
+                cross |= (ls.lat < latLonBounds.bottom - shift ? 8 : 0);
+                if (i > 0) {
+                    if ((prevCross & cross) == 0) {
+                        if (endIndex == i - 1 && startIndex != -1) {
+                            // continue previous line
+                        } else {
+                            // start new segment
+                            if (startIndex >= 0) {
+                                drawSegment(pts, p, canvas, tileBox, startIndex, endIndex);
+                            }
+                            startIndex = i - 1;
+                        }
+                        endIndex = i;
+                    }
+                }
+                prevCross = cross;
+            }
+            if (startIndex != -1) {
+                drawSegment(pts, p, canvas, tileBox, startIndex, endIndex);
+            }
+        }
+
+
+        private void drawSegment(List<GPXUtilities.WptPt> pts, Paint paint, Canvas canvas, RotatedTileBox tb, int startIndex, int endIndex) {
+            TIntArrayList tx = new TIntArrayList();
+            TIntArrayList ty = new TIntArrayList();
+            canvas.rotate(-tb.getRotate(), tb.getCenterPixelX(), tb.getCenterPixelY());
+            Path path = new Path();
+            for (int i = startIndex; i <= endIndex; i++) {
+                GPXUtilities.WptPt p = pts.get(i);
+                tx.add((int)(tb.getPixXFromLatLon(p.lat, p.lon) + 0.5));
+                ty.add((int)(tb.getPixYFromLatLon(p.lat, p.lon) + 0.5));
+            }
+
+//TODO: colour
```

Comment:
No, it isn’t :)
OK, the code there does a black border around tracks (another option). I think you may call this “shadow” but I’m not really sure - so it was just a note to indicate that that’s where the shadows go, and this should somehow be selectable via UI or variable.  This is something we need to resolve later.
> > +//TODO: colour
> Not clear what that todo means...
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub <https://github.com/osmandapp/Osmand/pull/2379/files/109ca692f100ddb59f0e538b06c32beae5910bc0#r57644732>